File: sections/69_master_structures_trees_mixed/problems/e_weighted_marked_pairs/solution.cpp
#include <bits/stdc++.h>
using namespace std;
using int64 = long long;
int main() {
  ios::sync_with_stdio(false);
  cin.tie(nullptr);
  int n, q;
  if (!(cin >> n >> q)) return 0;
  vector<vector<pair<int, int>>> g(n);
  for (int i = 1, u, v, w; i < n; i++) {
    cin >> u >> v >> w;
    --u;
    --v;
    g[u].push_back({v, w});
    g[v].push_back({u, w});
  }
  int log = bit_width((unsigned)n);
  vector<vector<int>> up(log, vector<int>(n));
  vector<int> tin(n), tout(n);
  vector<int64> distance(n);
  int timer = 0;
  vector<tuple<int, int, int>> stack{{0, 0, 0}};
  while (!stack.empty()) {
    auto [u, p, state] = stack.back();
    stack.pop_back();
    if (!state) {
      tin[u] = timer++;
      up[0][u] = p;
      for (int j = 1; j < log; j++) up[j][u] = up[j - 1][up[j - 1][u]];
      stack.push_back({u, p, 1});
      for (auto it = g[u].rbegin(); it != g[u].rend(); ++it)
        if (it->first != p)
          distance[it->first] = distance[u] + it->second,
          stack.push_back({it->first, u, 0});
    } else
      tout[u] = timer - 1;
  }
  auto ancestor = [&](int u, int v) { return tin[u] <= tin[v] && tin[v] <= tout[u]; };
  auto lca = [&](int u, int v) {
    if (ancestor(u, v)) return u;
    if (ancestor(v, u)) return v;
    for (int j = log - 1; j >= 0; j--)
      if (!ancestor(up[j][u], v)) u = up[j][u];
    return up[0][u];
  };
  vector<int64> weight(n), subtree(n);
  vector<int> parent(n);
  while (q--) {
    int k;
    cin >> k;
    vector<int> nodes(k);
    int64 total = 0;
    for (int i = 0; i < k; i++) {
      cin >> nodes[i] >> weight[--nodes[i]];
      total += weight[nodes[i]];
    }
    sort(nodes.begin(), nodes.end(), [&](int a, int b) { return tin[a] < tin[b]; });
    for (int i = 1; i < k; i++) nodes.push_back(lca(nodes[i - 1], nodes[i]));
    sort(nodes.begin(), nodes.end(), [&](int a, int b) { return tin[a] < tin[b]; });
    nodes.erase(unique(nodes.begin(), nodes.end()), nodes.end());
    vector<int> st;
    for (int u : nodes) {
      while (!st.empty() && !ancestor(st.back(), u)) st.pop_back();
      parent[u] = st.empty() ? -1 : st.back();
      st.push_back(u);
      subtree[u] = weight[u];
    }
    int64 answer = 0;
    for (int i = nodes.size() - 1; i >= 0; i--) {
      int u = nodes[i], p = parent[u];
      if (p != -1) {
        answer += (distance[u] - distance[p]) * subtree[u] * (total - subtree[u]);
        subtree[p] += subtree[u];
      }
    }
    cout << answer << '\n';
    for (int u : nodes) weight[u] = 0;
  }
}

File: sections/69_master_structures_trees_mixed/problems/e_weighted_marked_pairs/solution.py
import sys


def main():
    t = sys.stdin.buffer.read().split()
    if not t:
        return
    z = 0
    n = int(t[z])
    q = int(t[z + 1])
    z += 2
    g = [[] for _ in range(n)]
    for _ in range(n - 1):
        u = int(t[z]) - 1
        v = int(t[z + 1]) - 1
        w = int(t[z + 2])
        z += 3
        g[u].append((v, w))
        g[v].append((u, w))
    log = n.bit_length()
    up = [[0] * n for _ in range(log)]
    tin = [0] * n
    tout = [0] * n
    dist = [0] * n
    timer = 0
    stack = [(0, 0, 0)]
    while stack:
        u, p, state = stack.pop()
        if not state:
            tin[u] = timer
            timer += 1
            up[0][u] = p
            for j in range(1, log):
                up[j][u] = up[j - 1][up[j - 1][u]]
            stack.append((u, p, 1))
            for v, w in reversed(g[u]):
                if v != p:
                    dist[v] = dist[u] + w
                    stack.append((v, u, 0))
        else:
            tout[u] = timer - 1

    def anc(u, v):
        return tin[u] <= tin[v] <= tout[u]

    def lca(u, v):
        if anc(u, v):
            return u
        if anc(v, u):
            return v
        for j in range(log - 1, -1, -1):
            if not anc(up[j][u], v):
                u = up[j][u]
        return up[0][u]

    weight = [0] * n
    sub = [0] * n
    parent = [-1] * n
    out = []
    for _ in range(q):
        k = int(t[z])
        z += 1
        nodes = []
        total = 0
        for _ in range(k):
            u = int(t[z]) - 1
            c = int(t[z + 1])
            z += 2
            nodes.append(u)
            weight[u] = c
            total += c
        nodes.sort(key=tin.__getitem__)
        nodes += [lca(nodes[i - 1], nodes[i]) for i in range(1, k)]
        nodes = sorted(set(nodes), key=tin.__getitem__)
        st = []
        for u in nodes:
            while st and not anc(st[-1], u):
                st.pop()
            parent[u] = st[-1] if st else -1
            st.append(u)
            sub[u] = weight[u]
        answer = 0
        for u in reversed(nodes):
            p = parent[u]
            if p != -1:
                answer += (dist[u] - dist[p]) * sub[u] * (total - sub[u])
                sub[p] += sub[u]
        out.append(str(answer))
        for u in nodes:
            weight[u] = 0
    print("\n".join(out))


if __name__ == "__main__":
    main()
